### PR TITLE
AIOG-T-30 magic filter refactor

### PR DESCRIPTION
**Pull request by uwinx**
## [WIP] Refactored small part of codebase.

- ### typing
  - removed unecessarily added generic types and typevars
  - use `Iterable` instead of `Sequence` in `in_` or `__matmul__` (`@`)) operation (we could avoid translating value to set and keep them as `Container`). ref: [python collections docs](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes)

- ### attribute resolver
  - move resolver to attribute.py where it belongs

- ### BaseOperation and its inherited
  - rename run to "pythonic" `__call__`
  - removed `op` attribute 
  - added `__slots__` (some parts could be missed WIP)
  - remove dataclasses use
  - use tuple instead of list for modifiers

### Cython (rationale: tiny optimizations)
- added build.py reference in pyproject.toml

### CODE REVIEW

File: magic_filter/operations/bases.py [hunk 1]
```diff
@@ -1,67 +1,41 @@
...
         self.other_value = other_value
 
-    def run(self, obj: Any) -> bool:
-        return self.value.run(obj=obj) and self.other_value.run(obj=obj)
-
-    def __str__(self) -> str:
-        return f"({self.value}) {self.op} ({self.other_value})"
-
+    def __call__(self, obj: Any) -> Any:
+        return self.value.__call__(obj=obj) and self.other_value.__call__(obj=obj)
```
Comment (JrooTJunior) [inline magic_filter/operations/bases.py:31]:
```suggestion
        return self.value(obj=obj) and self.other_value(obj=obj)
```

File: magic_filter/operations/bases.py [hunk 2]
```diff
@@ -1,67 +1,41 @@
...
 
-class OrOperation(AndOperation[LeftOperand, RightOperand]):
-    op = "|"
 
-    def run(self, obj: Any) -> bool:
-        return self.value.run(obj) or self.other_value.run(obj)
+class OrOperation(AndOperation):
+    def __call__(self, obj: Any) -> Any:
+        return self.value.__call__(obj) or self.other_value.__call__(obj)
```
Comment (JrooTJunior) [inline magic_filter/operations/bases.py:36]:
```suggestion
        return self.value(obj) or self.other_value(obj)
```

File: magic_filter/operations/bases.py [hunk 3]
```diff
@@ -1,67 +1,41 @@
...
-    op = "~"
-
-    def run(self, obj: Any) -> bool:
-        return not self.value.run(obj)
-
-    def __str__(self) -> str:
-        return f"{self.op}({self.value})"
+class NotOperation(BaseOperation):
+    def __call__(self, obj: Any) -> bool:
+        return not self.value.__call__(obj)
```
Comment (JrooTJunior) [inline magic_filter/operations/bases.py:41]:
```suggestion
        return not self.value(obj)
```

File: magic_filter/magic.py [hunk 4]
```diff
@@ -103,23 +67,29 @@ def endswith(self, value: str) -> AttrOperation:
...
 
-    def any(self, *call: Callable[[Any, DataType], bool]) -> AnyFuncOperation:
+    def any(self, *call: Callable[[Any, Any], bool]) -> AnyFuncOperation:
         """
         Analog of builtin any(Iterable[Any])
         """
         return AnyFuncOperation(value=call, magic=self)
 
+    # PRIVATE API
+    def get_chain(self) -> List[Attribute]:
```
Comment (JrooTJunior) [inline magic_filter/magic.py:89]: What about property without setter?
```suggestion
    @property
    def chain(self) -> List[Attribute]:
```

File: magic_filter/magic.py [hunk 5]
```diff
@@ -103,23 +67,29 @@ def endswith(self, value: str) -> AttrOperation:
...
-    def func(self, func: Callable[..., Any]) -> FuncOperation:
+    def func(self, value: Callable[..., Any]) -> FuncOperation:
         """
         Execute any callable on value
         """
-        return FuncOperation(value=func, magic=self)
+        return FuncOperation(value=value, magic=self)
 
-    def all(self, *call: Callable[[Any, DataType], bool]) -> AllFuncOperation:
+    def all(self, *call: Callable[[Any, Any], bool]) -> AllFuncOperation:
```
Comment (JrooTJunior) [inline magic_filter/magic.py:76]: Data argument removed. In this case signature can be changed to:
```suggestion
    def all(self, *call: Callable[[Any], bool]) -> AllFuncOperation:
```

File: magic_filter/magic.py [hunk 6]
```diff
@@ -103,23 +67,29 @@ def endswith(self, value: str) -> AttrOperation:
...
         """
         Analog of builtin all(Iterable[Any])
         """
         return AllFuncOperation(value=call, magic=self)
 
-    def any(self, *call: Callable[[Any, DataType], bool]) -> AnyFuncOperation:
+    def any(self, *call: Callable[[Any, Any], bool]) -> AnyFuncOperation:
```
Comment (JrooTJunior) [inline magic_filter/magic.py:82]: Data argument removed. In this case signature can be changed to:
```suggestion
    def all(self, *call: Callable[[Any], bool]) -> AllFuncOperation:
```

File: magic_filter/operations/simple.py [hunk 7]
```diff
@@ -1,32 +1,24 @@
...
         self.magic = magic
 
-    @property
-    def chain(self) -> List[Attribute]:
-        return self.magic.chain
-
-    def run(self, obj: Any) -> bool:
-        value = self.magic.resolve_attribute(obj)
+    def __call__(self, obj: Any) -> bool:
+        value = resolve_attribute(obj, self.magic.get_chain())
```
Comment (JrooTJunior) [inline magic_filter/operations/simple.py:17]:
```suggestion
        value = resolve_attribute(obj, self.magic.chain)
```

File: magic_filter/magic.py (hunk 4 above)
Comment (uwinx) [reply magic_filter/magic.py:89]: clean namespace. object field may have chain attr. am I right?

Comment (JrooTJunior) [reply magic_filter/magic.py:89]: you're right
hmm...
what do you think about passing a chain instead of a filter instance in an operation?

Comment (uwinx) [reply magic_filter/magic.py:89]: I agree with this one

**Pull request merged**